Better password handling using pgpass

jonathang opened the pull request:
Inspired by #14

Seanstoppable commented on README.md:
```diff
@@ -212,6 +212,14 @@ You will likely see a number of create table statements (see data model section
+  sem-apply --url postgresql://postgres@localhost/sample --pgpass
+
+If your database server is protected by a password, you will be promped to enter the password on every sql command.
+To overcome this the aregumer `--pgpass` will generate a temporary .pgpass file which hold the password for the duration
```
Argument vs aregumer

Probably worth mentioning pgpass in general?

Seanstoppable commented on lib/schema-evolution-manager/apply_util.rb:
```diff
@@ -35,6 +36,24 @@ def apply!(dir)
+      puts "Creating temp pgpass at #{@pgpass_file.path}"
+      FileUtils.chmod(0600, @pgpass_file.path)
+
+      FileUtils.rm_rf("#{ENV['HOME']}/.pgpass")
```
This seems unintuitive for a user that you will nuke their pgpass file. There is a PGPASSWORD environment variable, which seems safer.

jonathang replied:
I agree it is a valid concern. 

I added a warning to the README, maybe I should also add a warning step before entering the password. I avoided going down the path of PGPASSWORD because of this comment:

http://www.postgresql.org/docs/9.1/static/libpq-envars.html

`PGPASSWORD behaves the same as the password connection parameter. Use of this environment variable is not recommended for security reasons, as some operating systems allow non-root users to see process environment variables via ps; instead consider using the ~/.pgpass file (see Section 31.14).`

jonathang replied:
Actually I can also just raise a run time exception if pgpass already exists 

Yoni

Sent from my mobile

> On Mar 16, 2016, at 6:59 PM, Sean Smith <notifications@github.com> wrote:
> 
> In lib/schema-evolution-manager/apply_util.rb:
> 
> > @@ -35,6 +36,24 @@ def apply!(dir)
> >        count
> >      end
> >  
> > +    def persist_pgpass(password)
> > +      Preconditions.check_not_blank(password, "password cannot be blank")
> > +      puts "Creating temp pgpass at #{@pgpass_file.path}"
> > +      FileUtils.chmod(0600, @pgpass_file.path)
> > +
> > +      FileUtils.rm_rf("#{ENV['HOME']}/.pgpass")
> This seems unintuitive for a user that you will nuke their pgpass file. There is a PGPASSWORD environment variable, which seems safer.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>

mbryzek replied:
Agreed - we should remove this - we'll never touch .pgpass

Seanstoppable commented on configure.rb:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env ruby
+a#!/usr/bin/env ruby
```
Accidental character

mbryzek commented on bin/sem-apply:
```diff
@@ -8,24 +8,33 @@
-
 dry_run = args.dry_run.nil? ? false : args.dry_run
+pgpass = args.pgpass.nil? ? false : args.pgpass
+pgpassfile = Tempfile.new('pgpass')
```
sem has a Library.with_temp_file helper that we should use here instead of tempfile

jonathang replied:
I am not sure why you are having Library.with_temp_file . Tempfile is doing exactly the same.

http://ruby-doc.org/stdlib-1.9.3/libdoc/tempfile/rdoc/Tempfile.html

```When a Tempfile object is garbage collected, or when the Ruby interpreter exits, its associated temporary file is automatically deleted.```

mbryzek replied:
thanks - I'm aware of temp file :)

two use cases I was interested in:
  1. I wanted a closure for all temp files
  2. when developing, I wanted an option to leave tmp files around for inspection

regardless - if we were to use Tempfile, we would switch everything to tempfile( removing library.with_temp_file). My preference is to keep Library.with_temp_file and thus not introduce Tempfile

mbryzek commented on README.md:
```diff
@@ -212,6 +212,19 @@ You will likely see a number of create table statements (see data model section
 
+### Database servers with passwords
+
+  sem-apply --url postgresql://postgres@localhost/sample --pgpass
```
is the intent to pass in "--pgpass <password>" ? or just a flag that then prompts for the password?

mbryzek replied:
also, I think we may want to consider renaming the parameter to "--password" - it actually has nothing to do with pgpass in psql - it is just the password.

mbryzek commented on bin/sem-apply:
```diff
@@ -8,24 +8,33 @@
+  if pgpass
+    puts "please enter your password (or press enter to continue)"
+    password = SchemaEvolutionManager::Ask.for_password()
+    if !(password.nil? || password.empty?)
```
I think if the password is empty, we should re-prompt until we get a password. ie. if I specify --pgpass, make me enter a password. Otherwise I can remove the --pgpass parameter

mbryzek commented on bin/sem-apply:
```diff
@@ -8,24 +8,33 @@
+
+  if pgpass
+    puts "please enter your password (or press enter to continue)"
+    password = SchemaEvolutionManager::Ask.for_password()
```
mind removing "()" ?

mbryzek commented on bin/sem-apply:
```diff
@@ -38,4 +47,8 @@ rescue SchemaEvolutionManager::ScriptError => e
   exit(1)
+ensure
+  if pgpass
+    util.destroy_pgpass()
```
This ensure block will be unnecessary with the change to Library.with_temp_file

jonathang replied:
it is best practice, it is also not needed with Tempfile.

```
it’s unnecessary to explicitly delete a Tempfile after use, though it’s good practice to do so
```

mbryzek commented on lib/schema-evolution-manager/apply_util.rb:
```diff
@@ -10,6 +10,7 @@ def initialize(db, opts={})
 
       @db = Preconditions.assert_class(db, Db)
       @scripts = Scripts.new(@db, Scripts::SCRIPTS)
+      @pgpass_file = Preconditions.assert_class(pgpass_file, Tempfile)
```
I think better is to make the argument of type string - the path to a file - rather than requiring a Tempfile

jonathang replied:
Currently I am writing to it, so unless we change it to the EV I need to pass the file

mbryzek commented on lib/schema-evolution-manager/apply_util.rb:
```diff
@@ -35,6 +36,24 @@ def apply!(dir)
       count
     end
 
+    def persist_pgpass(password)
```
Maybe for api, we change this to "with_password_file(password)"

this will make sure we can cleanup the file in one central place

jonathang replied:
WDYM - `this will make sure we can cleanup the file in one central place`?

mbryzek commented on lib/schema-evolution-manager/apply_util.rb:
```diff
@@ -35,6 +36,24 @@ def apply!(dir)
+      FileUtils.chmod(0600, @pgpass_file.path)
+
+      FileUtils.rm_rf("#{ENV['HOME']}/.pgpass")
+      FileUtils.ln_s(@pgpass_file.path, "#{ENV['HOME']}/.pgpass")
```
agree with sean's comments to use the PGPASSFILE env variable to point to our temp file

mbryzek commented on lib/schema-evolution-manager/apply_util.rb:
```diff
@@ -35,6 +36,24 @@ def apply!(dir)
+      @pgpass_file.rewind
+    end
+
+    def destroy_pgpass()
```
will not be needed with Library.with_temp_file

mbryzek commented on lib/schema-evolution-manager/args.rb:
```diff
@@ -19,13 +19,14 @@ class Args
     end
 
-    attr_reader :artifact_name, :host, :port, :name, :prefix, :url, :user, :dir, :dry_run, :tag
+    attr_reader :artifact_name, :host, :port, :name, :prefix, :url, :user, :dir, :dry_run, :tag, :pgpass
```
I don't think we should allow reading the password into ram... get it from STDIN and write to file then out of process

perl2ruby replied:
Since ".pgpass" file is automatically handled by the psql client should sem-apply simply fall back on  ".pgpass" being handled by the psql both to check the permissions (600 or more restrictive) and the correctness?

mbryzek replied:
yes - which btw is how things work now (e.g. you can manually create a .pgpass file and all works transparently with the psql client)

jonathang replied:
Removed. I just want it to be an empty argument. Like --dry_run.

mbryzek commented on lib/schema-evolution-manager/args.rb:
```diff
@@ -61,6 +62,7 @@ def initialize(args, opts={})
       @tag = found_arguments.delete(:tag)
 
       @dry_run = found_arguments.delete(:dry_run)
+      @pgpass = found_arguments.delete(:pgpass)
```
Would recommend replacing with pg_pass_file

mbryzek commented:
Love this change! Thanks for taking the time to put together the PR. This is going to be incredibly useful.

I think main pieces for us all to align on are param naming, making sure we never touch ~/.pgass, and maybe some more testing around going from the connection url to the pgpass file format.

jonathang commented:
Thanks for the prompt review and for the great feedback.

I think that I have addressed most of your concerns. Two open points:

- I think that the updated approach on .pgpass is more reliable (won't nuke it anymore) and recommended (based on postgres docs) than using the EV. But if both of you  (@mbryzek and @Seanstoppable) still think that I should go down the EV path I will make the change.

- I am not sure why are you using the helper in Library.with_temp_file  Ruby's Tempfile seems to do the 
same. I actually think that it be cleaner to replace the helper with Ruby's Tempfile.

Thoughts?

jonathang commented:
Ok my last change is using the PGPASSFILE (not PGPASSWORD) and it doesn't touch at all the ~/.pgpass

@mbryzek - what else do you want me to address?

Seanstoppable commented on README.md:
```diff
@@ -212,6 +212,22 @@ You will likely see a number of create table statements (see data model section
+.pgpass file which will hold the password for the duration of the sem-apply.
+
+This argument will create a temp `.pgpass` file with your postgres server password, which will be deleted upon exiting the
+Ruby interpreter. On the other hand, if you are not concerend about placing the password in `~/.pgpass` , you can do that
```
concerned

Seanstoppable commented on README.md:
```diff
@@ -212,6 +212,22 @@ You will likely see a number of create table statements (see data model section
+
+  sem-apply --url postgresql://postgres@localhost/sample --password
+
+If your database server is protected by a password, you will be promped to enter the password on every sql command.
```
prompted

Seanstoppable commented on README.md:
```diff
@@ -212,6 +212,22 @@ You will likely see a number of create table statements (see data model section
+  sem-apply --url postgresql://postgres@localhost/sample --password
+
+If your database server is protected by a password, you will be promped to enter the password on every sql command.
+To overcome this the argument `--password` will prompt you to enter your password and it will generate a temporary
```
"enter your password once and then"

Seanstoppable commented on README.md:
```diff
@@ -212,6 +212,22 @@ You will likely see a number of create table statements (see data model section
+Ruby interpreter. On the other hand, if you are not concerend about placing the password in `~/.pgpass` , you can do that
+and then you won't need to use the  `--password` attribute.
+
+[To read more about `.pgpass`](http://www.postgresql.org/docs/9.4/static/libpq-pgpass.html)
```
:+1:

Seanstoppable commented:
Yeah, my bad, I was trying to refer to PGPASSFILE. Seems like the best way to go here. :+1: 

Some minor README language comments. I will leave you and @mbryzek to discuss the pros and cons about using tempfile.

mbryzek commented:
I've taken a pass at refactoring:

https://github.com/mbryzek/schema-evolution-manager/pull/34

While many of the changes are largely subjective - this is the general style I would like to proceed in.

I've installed locally and would like to test locally over the next few days before releasing.

comments welcome
